src/main/java/hudson/remoting/AbstractByteBufferCommandTransport.java

```diff
@@ -0,0 +1,286 @@
+                    }
+                    break;
+            }
+        }
+    }
+
+    private void processCommand() throws IOException {
+        try {
+            FastByteBufferQueueInputStream is = new FastByteBufferQueueInputStream(receiveQueue, readCommandSizes[0]);
+            try {
+                ObjectInputStreamEx ois = new ObjectInputStreamEx(is, channel.baseClassLoader, channel.classFilter);
+                receiver.handle(Command.readFrom(channel, ois));
+            } catch (IOException e1) {
+                LOGGER.log(Level.WARNING, "Failed to construct Command", e1);
+            } catch (ClassNotFoundException e11) {
+                LOGGER.log(Level.WARNING, "Failed to construct Command", e11);
+            } finally {
+                int available = is.available();
+                if (available > 0) {
+                    is.skip(available);
+                }
+                IOUtils.closeQuietly(is);
+            }
+        } finally {
+            if (readCommandIndex == 1) {
+                if (readCommandSizes.length > 16) {
+                    // de-allocate the grown array
+                    int temp = readCommandSizes[1];
+                    readCommandSizes = new int[16];
+                    readCommandSizes[0] = temp;
+                } else {
+                    readCommandSizes[0] = readCommandSizes[1];
+                }
+                readCommandIndex = 0;
+            } else {
+                // we have processed position 0
+                // we now want position 1 to be the new position 0
+                // position readCommandIndex has the readState data length, so need to keep that
+                System.arraycopy(readCommandSizes, 1, readCommandSizes, 0, readCommandIndex);
+                readCommandIndex--;
+            }
+        }
+    }
+
+    /**
+     * Set the frame size.
+     *
+     * @param transportFrameSize the new frame size.
+     */
+    public void setFrameSize(int transportFrameSize) {
+        if (transportFrameSize <= 0 || transportFrameSize >= Short.MAX_VALUE) {
+            throw new IllegalArgumentException();
+        }
+        this.transportFrameSize = transportFrameSize;
+        // this is the only one that matters when it comes to sizing as we have to accept any frame size on receive
+        writeChunkBody = ByteBuffer.allocate(transportFrameSize);
+    }
+
+
+    @Override
+    public final synchronized void setup(final Channel channel, final CommandReceiver receiver) {
+        this.channel = channel;
+        this.receiver = receiver;
+        try {
+            // just in case any data was queued while we were waiting for the call to setup, let's trigger the
+            // flush of the receive queue.
+            while (receiver != null && readCommandIndex > 0) {
+                processCommand();
+            }
+        } catch (IOException e) {
+            LOGGER.log(Level.WARNING, "Could not flush receive buffer queue", e);
```

Comment:
Nope that's the way remoting works... in part because other command transports are not able to ensure that they stay synced